# Guided Capstone Step 4 - Pre-processing and Training Data Development

### **The Data Science Method**  


1.   Problem Identification 

2.   Data Wrangling 
  
3.   Exploratory Data Analysis 
 
4.   **Pre-processing and Training Data Development**

5.  Modeling
  * Fit Models with Training Data Set
  * Review Model Outcomes — Iterate over additional models as needed.
  * Identify the Final Model

6.   Documentation
  * Review the Results
  * Present and share your findings - storytelling
  * Finalize Code 
  * Finalize Documentation

In this step of the guided capstone, you'll revisit some of the steps you completed to solve questions in step 4. Revisiting these steps will help you get additional practice completing this kind of work.

**<font color='teal'> Start by loading the necessary packages, as you've done in the previous steps. Print out your current working directory to confirm that you are in the correct project directory. </font>**

In [2]:
#load python packages
import os
import pandas as pd
import datetime
import seaborn as sns
import matplotlib.pyplot as plt
import numpy as np
%matplotlib inline
os.getcwd()

'/Users/lisahw/Documents/Courses and Conferences/DataScience/MyProject/Unit 6'

**<font color='teal'>  Load the csv file you created in step 3. Remember, it should be saved inside your data subfolder. Print the first five rows.</font>**

In [45]:
df = pd.read_csv('data/step3_output_temp.csv',index_col=0)
df.head()

Name    state  summit_elev  vertical_drop  trams  fastSixes  \
0       Alyeska Resort   Alaska       3939.0         2500.0      1          0   
1  Eaglecrest Ski Area   Alaska       2600.0         1540.0      0          0   
2     Hilltop Ski Area   Alaska       2090.0          294.0      0          0   
3     Arizona Snowbowl  Arizona      11500.0         2300.0      0          1   
4  Sunrise Park Resort  Arizona      11100.0         1800.0      0          0   

   fastQuads  quad  triple  double  ...  LongestRun_mi  SkiableTerrain_ac  \
0          2     2       0       0  ...            1.0             1610.0   
1          0     0       0       4  ...            2.0              640.0   
2          0     0       1       0  ...            1.0               30.0   
3          0     2       2       1  ...            2.0              777.0   
4          1     2       3       1  ...            1.2              800.0   

   Snow Making_ac  daysOpenLastYear  yearsOpen  averageSnowfall  AdultWeekday  \
0           113.0        150.000000       60.0       183.203145          65.0   
1            60.0        115.103943       44.0       350.000000          47.0   
2            30.0        150.000000       36.0        69.000000          30.0   
3           104.0        122.000000       81.0       260.000000          89.0   
4            80.0        115.000000       49.0       250.000000          74.0   

   AdultWeekend  projectedDaysOpen  NightSkiing_ac  
0          85.0              150.0       56.890909  
1          53.0               90.0        0.000000  
2          34.0              152.0       30.000000  
3          89.0              122.0        0.000000  
4          78.0              104.0       80.000000  

[5 rows x 24 columns]

## Create dummy features for categorical variables

**<font color='teal'> Create dummy variables for `State`. Add the dummies back to the dataframe and remove the original column for `State`. </font>**

In [46]:
# dfo= df.select_dtypes(include=['object']) # select object type columns
dft = pd.concat([df.drop('state', axis=1), pd.get_dummies(df['state'])], axis=1)
dft.head()

Name  summit_elev  vertical_drop  trams  fastSixes  \
0       Alyeska Resort       3939.0         2500.0      1          0   
1  Eaglecrest Ski Area       2600.0         1540.0      0          0   
2     Hilltop Ski Area       2090.0          294.0      0          0   
3     Arizona Snowbowl      11500.0         2300.0      0          1   
4  Sunrise Park Resort      11100.0         1800.0      0          0   

   fastQuads  quad  triple  double  surface  ...  Rhode Island  South Dakota  \
0          2     2       0       0        2  ...             0             0   
1          0     0       0       4        0  ...             0             0   
2          0     0       1       0        2  ...             0             0   
3          0     2       2       1        2  ...             0             0   
4          1     2       3       1        0  ...             0             0   

   Tennessee  Utah  Vermont  Virginia  Washington  West Virginia  Wisconsin  \
0          0     0        0         0           0              0          0   
1          0     0        0         0           0              0          0   
2          0     0        0         0           0              0          0   
3          0     0        0         0           0              0          0   
4          0     0        0         0           0              0          0   

   Wyoming  
0        0  
1        0  
2        0  
3        0  
4        0  

[5 rows x 58 columns]

## Standardize the magnitude of numeric features

**<font color='teal'> Using sklearn preprocessing, standardize the scale of the features of the dataframe except the name of the resort, which you don't need in the dataframe for modeling so it can be droppped here as well. You should  hold out your response variable(s) so that you can have their true values available for model performance review. Let's set `AdultWeekend` to the y variable as your response for scaling and modeling. Later, you will go back and consider the `AdultWeekday`, `dayOpenLastYear`, and `projectedDaysOpen`. For now leave, them in the development dataframe. </font>**

In [14]:
np.where(dft.columns=='AdultWeekend')

(array([19]),)

In [47]:
from sklearn.preprocessing import StandardScaler

dft.drop('Name',axis=1,inplace=True)
scaler = StandardScaler()
scaler.fit(dft.values)
print(scaler.mean_)
dfn = scaler.transform(dft.values)
# dfn = scaler.fit_transform(dft.values)
# dfn = dft.AdultWeekend
y_aw = dfn[:,19]

[4.59181818e+03 1.17767488e+03 1.72727273e-01 1.84848485e-01
 1.01818182e+00 9.33333333e-01 1.50000000e+00 1.83333333e+00
 2.62121212e+00 7.20404040e+00 4.03695111e+01 2.32059303e+00
 1.31817716e+00 3.89283792e+02 1.29131797e+02 1.13928109e+02
 5.98257097e+01 1.80277104e+02 5.45737879e+01 6.11708565e+01
 1.19924146e+02 4.27805510e+01 9.09090909e-03 6.06060606e-03
 6.36363636e-02 6.66666667e-02 1.51515152e-02 3.63636364e-02
 1.21212121e-02 6.06060606e-03 9.09090909e-03 2.72727273e-02
 3.03030303e-03 3.33333333e-02 8.78787879e-02 4.24242424e-02
 6.06060606e-03 3.63636364e-02 1.21212121e-02 4.84848485e-02
 6.06060606e-03 2.72727273e-02 1.00000000e-01 1.81818182e-02
 1.51515152e-02 3.03030303e-02 5.75757576e-02 3.03030303e-03
 6.06060606e-03 3.03030303e-03 3.93939394e-02 4.54545455e-02
 1.21212121e-02 3.03030303e-02 1.21212121e-02 4.84848485e-02
 2.42424242e-02]


In [21]:
dft['AdultWeekend'].head(20)

0     85.00000
1     53.00000
2     34.00000
3     89.00000
4     78.00000
5     47.00000
6     64.16681
7     64.16681
8     64.16681
9     78.00000
10    75.00000
11    64.16681
12    64.16681
13    64.16681
14    64.16681
15    59.00000
16    84.00000
17    69.00000
18    64.16681
19    64.16681
Name: AdultWeekend, dtype: float64

## Split into training and testing datasets

**<font color='teal'> Using sklearn model selectionm import train_test_split, and create a 75/25 split with the y = `AdultWeekend`. You will start by using the adult weekend ticket price as your response variable for modeling.</font>**

In [48]:
from sklearn.model_selection import train_test_split 
dfn_train, dfn_test = train_test_split(dfn,test_size=0.25)

In [49]:
print(dfn_train.shape,dfn_test.shape)

(247, 57) (83, 57)


In [50]:
dfn = pd.DataFrame(dfn,columns=dft.columns)
dfn.head()

summit_elev  vertical_drop     trams  fastSixes  fastQuads      quad  \
0    -0.175024       1.500551  1.479658  -0.284078   0.447306  0.814091   
1    -0.534018       0.411160 -0.308940  -0.284078  -0.463873 -0.712330   
2    -0.670752      -1.002778 -0.308940  -0.284078  -0.463873 -0.712330   
3     1.852123       1.273594 -0.308940   1.252736  -0.463873  0.814091   
4     1.744881       0.706203 -0.308940  -0.284078  -0.008283  0.814091   

     triple    double   surface  total_chairs  ...  Rhode Island  \
0 -0.927831 -1.011619 -0.302071     -0.061876  ...     -0.055132   
1 -0.927831  1.195550 -1.274590     -0.971645  ...     -0.055132   
2 -0.309277 -1.011619 -0.302071     -1.274901  ...     -0.055132   
3  0.309277 -0.459827 -0.302071      0.241380  ...     -0.055132   
4  0.927831 -0.459827 -1.274590     -0.061876  ...     -0.055132   

   South Dakota  Tennessee      Utah   Vermont  Virginia  Washington  \
0     -0.078087  -0.055132 -0.202508 -0.218218  -0.11077   -0.176777   
1     -0.078087  -0.055132 -0.202508 -0.218218  -0.11077   -0.176777   
2     -0.078087  -0.055132 -0.202508 -0.218218  -0.11077   -0.176777   
3     -0.078087  -0.055132 -0.202508 -0.218218  -0.11077   -0.176777   
4     -0.078087  -0.055132 -0.202508 -0.218218  -0.11077   -0.176777   

   West Virginia  Wisconsin   Wyoming  
0       -0.11077  -0.225733 -0.157622  
1       -0.11077  -0.225733 -0.157622  
2       -0.11077  -0.225733 -0.157622  
3       -0.11077  -0.225733 -0.157622  
4       -0.11077  -0.225733 -0.157622  

[5 rows x 57 columns]

Here, you'll start the actual modeling work. First let's fit a multiple linear regression model to predict the `AdultWeekend` price.

# Guided Capstone Step 5 - Modeling

You've cleaned and prepared the datasets, so now it's time to get into the most exciting part of this process: modeling! In this exercise, you will build three different models and compare each model's performance. In the end, you will choose the best model for proving insights to Big Mountain management.  

1.   Problem Identification 

2.   Data Wrangling 
  
3.   Exploratory Data Analysis 
 
4.   Pre-processing and Training Data Development

5.   **Modeling**
  * Fit Models with Training Data Set
  * Review Model Outcomes — Iterate over additional models as needed.
  * Identify the Final Model

6.   Documentation
  * Review the Results
  * Present and share your findings - storytelling
  * Finalize Code 
  * Finalize Documentation

## Fit Models with Training Data Set

**<font color='teal'> Using sklearn, fit the model on your training dataset.</font>**

#### Model 1

In [141]:
X_train = np.delete(dfn_train,19,1)
y_train = dfn_train[:,19]
X_test = np.delete(dfn_test,19,1)
y_test = dfn_test[:,19]

In [142]:
# all first model set
from sklearn import linear_model
from sklearn.metrics import explained_variance_score,mean_absolute_error
lm = linear_model.LinearRegression()
model = lm.fit(X_train,y_train)

**<font color='teal'> Predict on the testing dataset and score the model performance with the y_test set and the y-pred values. The explained variance is a measure of the variation explained by the model. This is also known as the R-squared value. </font>**

In [143]:
model.coef_

array([ 1.20970256e-01,  8.00471008e-02, -2.45340861e-02,  1.30060002e-02,
       -1.08189545e-02,  7.32374820e-02,  5.88694691e-02,  3.43045102e-02,
       -5.96019042e-02,  7.73666980e-02,  2.27864435e-02, -3.38872639e-04,
        1.22922818e-03, -1.09644539e-02, -1.19488910e-03,  1.30671690e-02,
       -3.94676695e-03,  5.06446991e-02,  7.19035241e-01, -8.40994405e-04,
        8.66343972e-02, -1.86442765e+10, -1.52462470e+10, -4.79512485e+10,
       -4.90001863e+10, -2.39959371e+10, -3.67718276e+10, -2.14956127e+10,
       -1.52462470e+10, -1.86442765e+10, -3.19951981e+10, -1.07971462e+10,
       -3.52616570e+10, -5.56151713e+10, -3.95930166e+10, -1.52462470e+10,
       -3.67718276e+10, -2.14956127e+10, -4.21925572e+10, -1.52462470e+10,
       -3.19951981e+10, -5.89313279e+10, -2.62457606e+10, -2.39959371e+10,
       -3.36733265e+10, -4.57581047e+10, -1.07971462e+10, -1.52462470e+10,
       -1.07971462e+10, -3.82131061e+10, -4.09177683e+10, -2.14956127e+10,
       -3.36733265e+10, -

In [144]:
from sklearn.metrics import mean_squared_error, r2_score
y_pred = model.predict(X_test)
print('Explained Variance: ',explained_variance_score(y_test,y_pred))
print('Mean square error: ',mean_squared_error(y_test,y_pred))
print('R2 score/Coeff. of determination: ',r2_score(y_test,y_pred))

Explained Variance:  0.8290994834812799
Mean square error:  0.1730337674598858
R2 score/Coeff. of determination:  0.826889393588716


## Review Model Outcomes — Iterate over additional models as needed.

In [62]:
from sklearn.ensemble import RandomForestRegressor
regressor = RandomForestRegressor(random_state=0, n_estimators=200)
regressor.fit(X_train, y_train)
regressor.score(X_test, y_test)

0.8042202660337793

In [63]:
y_pred = regressor.predict(X_test)
print('Mean square error: ',mean_squared_error(y_test,y_pred))
print('R2 score/Coeff. of determination: ',r2_score(y_test,y_pred))

Mean square error:  0.1956928328237961
R2 score/Coeff. of determination:  0.8042202660337792


In [64]:
from sklearn.model_selection import GridSearchCV
param_grid = {"n_estimators": [200, 500],
 "max_depth": [3, None],
 "max_features": [1, 3, 5, 10],
 "min_samples_split": [2, 5, 10],
 "min_samples_leaf": [1, 3, 10],
 "bootstrap": [True, False]}
reg2 = RandomForestRegressor(random_state=0)
grid = GridSearchCV(estimator=reg2, param_grid=param_grid, n_jobs=-1)
grid.fit(X_train, y_train)
print(grid.best_score_)
print(grid.best_params_)

0.7153975628225474
{'bootstrap': False, 'max_depth': None, 'max_features': 10, 'min_samples_leaf': 1, 'min_samples_split': 2, 'n_estimators': 500}


### WHY IS THE BEST SCORE WORSE THAN WHAT WE HAD PREVIOUSLY?

**<font color='teal'> Print the intercept value from the linear model. </font>**

In [89]:
print(model.intercept_)
coeff = pd.DataFrame(model.coef_,index=dfn.columns.drop('AdultWeekend'))
coeff.colums = ['Coef']

0.011796950709900486


**<font color='teal'> The intercept is the mean `AdultWeekend` price for all the resorts given the other characteristics. The addition or subtraction of each of the coefficient values in the regression are numeric adjustments applied to the intercept to provide a particular observation's value for the `AdultWeekend` resulting value. Also, because you took the time to scale your x values in the training data, you can compare each of the coefficients for the features to determine the feature's importance. Print the coefficient values from the linear model and sort in descending order to identify the top ten most important features.</font>** 

Hint: make sure to review the absolute value of the coefficients, because the adjustment may be positive or negative, but what you are looking for is the magnitude of impact on your response variable.

In [101]:
coeff.apply(abs).sort_values(by='Coef',axis=0,ascending=False).head(10)

Coef
New York       5.893133e+10
Michigan       5.561517e+10
Colorado       4.900019e+10
California     4.795125e+10
Pennsylvania   4.575810e+10
Wisconsin      4.219256e+10
New Hampshire  4.219256e+10
Vermont        4.091777e+10
Minnesota      3.959302e+10
Utah           3.821311e+10

**<font color='teal'>You should see that the top ten important features are different states. However, the state is not something the managers at the Big Mountain Resort can do anything about. Given that you care more about actionable traits associated with ticket pricing, rebuild the model without the state features and compare the results. </font>**

#### Model 2

In [104]:
print(np.where(dfn.columns=='Alaska'))
print(len(dfn.columns))

(array([22]),)
57


In [145]:
temp = np.delete(dfn_train,range(22,57),1)
X_train = np.delete(temp,19,1)
y_train = temp[:,19]
temp = np.delete(dfn_test,range(22,57),1)
X_test = np.delete(temp,19,1)
y_test = temp[:,19]

In [146]:
lm = linear_model.LinearRegression()
model = lm.fit(X_train,y_train)
# Predict AdultWeekend with test data
y_pred = model.predict(X_test)
print('Explained Variance: ',explained_variance_score(y_test,y_pred))
print('Mean square error: ',mean_squared_error(y_test,y_pred))
print('R2 score/Coeff. of determination: ',r2_score(y_test,y_pred))
# Analyze Intercept
print(model.intercept_)
coeff = pd.DataFrame(model.coef_,index=dfn.columns[:22].drop('AdultWeekend'))

Explained Variance:  0.7911331829721059
Mean square error:  0.20902619635560116
R2 score/Coeff. of determination:  0.7908809815670754
0.0039933878880368335


In [121]:
coeff.columns = ['Coef']
coeff.apply(abs).sort_values(by='Coef',axis=0,ascending=False).head(10)

Coef
AdultWeekday      0.738998
vertical_drop     0.201036
total_chairs      0.093527
summit_elev       0.090695
quad              0.086219
averageSnowfall   0.081388
NightSkiing_ac    0.077864
surface           0.076613
daysOpenLastYear  0.049108
TerrainParks      0.048219

**<font color='teal'> When reviewing your new model coefficients, note that `summit_elev` is now in the number two spot. From a managerial perspective, this is also difficult to change and highly correlated with `base_elev` and `vertical_drop`.  This time rebuild the model without the state features and without the `summit_elev` and without `base_elev`and compare the results. </font>**

#### Model 3

In [147]:
df_train = pd.DataFrame(dfn_train,columns=dfn.columns)
df_test = pd.DataFrame(dfn_test,columns=dfn.columns)

In [148]:
df_train.drop(['summit_elev','vertical_drop','Alaska',
       'Arizona', 'California', 'Colorado', 'Connecticut', 'Idaho', 'Illinois',
       'Indiana', 'Iowa', 'Maine', 'Maryland', 'Massachusetts', 'Michigan',
       'Minnesota', 'Missouri', 'Montana', 'Nevada', 'New Hampshire',
       'New Jersey', 'New Mexico', 'New York', 'North Carolina', 'Ohio',
       'Oregon', 'Pennsylvania', 'Rhode Island', 'South Dakota', 'Tennessee',
       'Utah', 'Vermont', 'Virginia', 'Washington', 'West Virginia',
       'Wisconsin','Wyoming'],axis=1,inplace=True)
df_test.drop(['summit_elev','vertical_drop','Alaska',
       'Arizona', 'California', 'Colorado', 'Connecticut', 'Idaho', 'Illinois',
       'Indiana', 'Iowa', 'Maine', 'Maryland', 'Massachusetts', 'Michigan',
       'Minnesota', 'Missouri', 'Montana', 'Nevada', 'New Hampshire',
       'New Jersey', 'New Mexico', 'New York', 'North Carolina', 'Ohio',
       'Oregon', 'Pennsylvania', 'Rhode Island', 'South Dakota', 'Tennessee',
       'Utah', 'Vermont', 'Virginia', 'Washington', 'West Virginia',
       'Wisconsin','Wyoming'],axis=1,inplace=True)

In [132]:
df_train.columns

Index(['trams', 'fastSixes', 'fastQuads', 'quad', 'triple', 'double',
       'surface', 'total_chairs', 'Runs', 'TerrainParks', 'LongestRun_mi',
       'SkiableTerrain_ac', 'Snow Making_ac', 'daysOpenLastYear', 'yearsOpen',
       'averageSnowfall', 'AdultWeekday', 'AdultWeekend', 'projectedDaysOpen',
       'NightSkiing_ac'],
      dtype='object')

In [149]:
X_train = df_train.drop('AdultWeekend',axis=1).values
y_train = df_train.AdultWeekend.values
X_test = df_test.drop('AdultWeekend',axis=1).values
y_test = df_test.AdultWeekend.values

In [150]:
lm = linear_model.LinearRegression()
model = lm.fit(X_train,y_train)
# Predict AdultWeekend with test data
y_pred = model.predict(X_test)
print('Explained Variance: ',explained_variance_score(y_test,y_pred))
print('Mean square error: ',mean_squared_error(y_test,y_pred))
print('R2 score/Coeff. of determination: ',r2_score(y_test,y_pred))
# Analyze Intercept
print(model.intercept_)
coeff = pd.DataFrame(model.coef_,index=df_train.columns.drop('AdultWeekend'))

Explained Variance:  0.7997730650447745
Mean square error:  0.20054859999533034
R2 score/Coeff. of determination:  0.7993623425660308
0.005096147408552065


In [140]:
coeff.columns = ['Coef']
coeff.apply(abs).sort_values(by='Coef',axis=0,ascending=False).head(10)

Coef
AdultWeekday      0.766066
averageSnowfall   0.099046
total_chairs      0.095148
quad              0.083681
surface           0.080780
NightSkiing_ac    0.071284
LongestRun_mi     0.066687
daysOpenLastYear  0.044751
TerrainParks      0.040134
fastSixes         0.038280

## Identify the Final Model

**<font color='teal'> Review the model performances in the table below and choose the best model for proving insights to Big Mountain management about what features are driving ski resort lift ticket prices. Type your choice in the final markdown cell. You will explain your selection during the next step of the guided casptone. </font>**

| Model | Explained Variance| Mean Absolute Error|Features Dropped|
| --- | --- | --- | --- |
| Model 1. | 0.829 | 0.173  |-|
| Model 2. | 0.791| 0.209 |'state'|
| Model 3. | 0.799 | 0.200|'state','summit_elev','base_elev'|

Model Selection:

There appears to be a geographical/political influence on the lift ticket prices with a dependency on the given State. Still, since the management cannot move the Big Mountain Resort away from Montana, I would focus the analysis on **Model 3** which excludes static factors.